experimental/dispatch_profiler/library.py

```diff
@@ -0,0 +1,295 @@
+# are built using the above enumerations. These data structures are used to create compose bigger
+# data structures that describe an operation or a sequence of operations, along with compilation
+# pipeling to form a collection of dispatches to profiled.
+###################################################################################################
+
+
+class TensorDescription:
+  """A class for tensor description."""
+
+  def __init__(self, datatype, layout):
+    self.datatype = datatype
+    self.layout = layout
+
+  def name(self):
+    return "%s%s" % (DataTypeName[self.datatype],
+                     ShortLayoutTypeName[self.layout])
+
+
+class TileDescription:
+  """A class for tile description."""
+
+  def __init__(self, threadblock_shape, stages, block_dim):
+    self.threadblock_shape = threadblock_shape  # in number of elements in M, N, K
+    self.stages = stages  # number of shared memory stages in tile K
+    self.block_dim = block_dim  # block dimension in number of threads in x, y, z
+
+  def name(self):
+    return "%dx%d_%dx%d" % (self.threadblock_shape[0],
+                            self.threadblock_shape[1],
+                            self.threadblock_shape[2], self.stages)
+
+
+###################################################################################################
+# The following part contains utility functions for which are used by the profiler tool.
+# These function may be moved as the need for create a proper structure for the
+# functionality they provide becomes apparent and necessary as we move forward.
+###################################################################################################
+# Helper function to generate a npy file name from a distribution and shape.
+def get_np_array(tensor_description, shape, dist):
+  if dist == Distribution.Empty:
+    return np.empty(shape)
+  elif dist == Distribution.Zeros:
+    return np.zeros(shape)
+  elif dist == Distribution.Ones:
+    return np.ones(shape)
+  elif dist == Distribution.Sequential:
+    return np.arange(np.prod(shape)).reshape(shape)
+  elif dist == Distribution.Identity:
+    return np.eye(shape[0], shape[1])
+  elif dist == Distribution.Random:
+    if tensor_description.datatype == DataType.s8:
+      return np.random.randint(-2, 2, shape)
+    elif tensor_description.datatype == DataType.u8:
+      return np.random.randint(0, 4, shape)
+    elif tensor_description.datatype == DataType.f16 or \
+         tensor_description.datatype == DataType.bf16:
+      return np.random.randint(-4, 4, shape)
+    elif tensor_description.datatype == DataType.f32:
+      return np.random.randint(-8, 8, shape)
+
+
+###################################################################################################
+# Helper function to substitute values into a template string.
+def SubstituteTemplate(template, values):
+  text = template
+  changed = True
+  while changed:
```

Comment:
What's the reason we need to do iterative substitution here?

Reply:
It is more robust and more use-cases, but may not be needed here, so simplified it for now. Thanks for going deep into this part.